### PR TITLE
Clip semantic search

### PR DESCRIPTION
**Pull request by sugi-chan**
Creating initial class to handle searching with CLIP + Faiss.  Hopefully this helps with dataset building type things.

Pull Request Checklist
 - [x] Pull request includes a description of why we are doing this
 - [x] `CHANGELOG.md` has been updated
 - [x] Version in `_version.py` has been updated
 - [x] All tests in the `tests` folder pass with a local build
 - [x] `requirements.txt` and `requirements-dev.txt` have been recompiled (if applicable)
 - [x] `__init__.py` file(s) import new capabilities to appropriate level of package (if applicable)
 - [x] `README.md` has been updated (if applicable)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,24 +28,20 @@ Now the library wheel will be uploaded to Databricks FileStore and can be instal
...
-Note: this will only work within Databricks for installation.
+clone repo and install
 
 ## Development
 
 ### Local development
-Explain how to add to development of this library, specific Databricks resources? Docker?
-
+clone repo
+cd into dir. install as editable library
```
Comment (sugi-chan) [inline README.md:41]: i should have actual instructions for a lot of the readme.... but I will probably make it another ticket/PR since I need to make and add the logo as well and such

File: fairywren/clip/faiss_index_search.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+
+        input_dict = [{'text':'hello world','image':SOME_PIL_IMAGE},
+                      {'text':'hello world2','image':SOME_PIL_IMAGE2}]
+
+        input_dict = [{'text':'hello world','image':SOME_PIL_IMAGE},
+                      {'image':SOME_PIL_IMAGE2}]
+        ```
+
+        PARAMETERS
+        ----------
```
Comment (emmafgrimes) [inline fairywren/clip/faiss_index_search.py:133]: Should we include `max_length` here?

File: fairywren/clip/faiss_index_search.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        Takes the output of a CLIP preprocessor and runs it through the CLIP model
+        depending on the type of data it will use a different extraction function.
+
+        Samples with only images OR text will use the `get_X_features` method of the
+        huggingface CLIP models and if there are both images and text then it will use
+        the base model forward pass. The model forward pass returns an array of data which
+        is more than just the embeddings. Thus the user will likely have to parse it themselves
+
+        If a user uses both images and text the features can be extracted with
+        `image_embeds` and `text_embeds` fields in the CLIP output. Fairywern is focused on
```
Comment (emmafgrimes) [inline fairywren/clip/faiss_index_search.py:196]: Typo
```suggestion
        `image_embeds` and `text_embeds` fields in the CLIP output. Fairywren is focused on
```

File: fairywren/clip/faiss_index_search.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+        image OR text vs faiss index. At the moment which means it is up to the user to do searches
+        with non standard length 512 embeddings.
+
+        Parameters
+        ----------
+            clip_inputs: dict
+                a dictionary with keys depending on the type of data that was preprocessed.
+                if there are images it will contain a `pixel_values` key with a tensor image/s
+                as the value and if there is text it will contain`input_ids` and `attention_mask`
+                keys and values.
```
Comment (emmafgrimes) [inline fairywren/clip/faiss_index_search.py:206]: [nitpicky] can we fix the indentation here and below?

File: fairywren/clip/faiss_index_search.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+                    **clip_inputs)
+
+            elif text_inputs and image_inputs:
+                embeddings = self.extraction_clip_model(**clip_inputs)
+
+            return embeddings
+
+    def search(self, embedding_array, result_count=10):
+        """
+        Takes and embedding or array of embeddings, searches the built Faiss index, adds metadata,
```
Comment (emmafgrimes) [inline fairywren/clip/faiss_index_search.py:237]: Typo
```suggestion
        Takes an embedding or array of embeddings, searches the built Faiss index, adds metadata,
```

File: fairywren/clip/faiss_index_search.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+        result_count: int
+            number of results to return, defaults to 10
+
+        Returns
+        -------
+            list of dictionaries with 3 keys. `metadata`, `dist`, and `indexes`
+                metadata: contains metadata for each item returned by the faiss index search
+                    based on the metadata parameter when the CLIPIndex class is built
+                dist: The reported distance between seed embedding and results.
+                I: the index numbers of results from the faiss index search.
```
Comment (emmafgrimes) [inline fairywren/clip/faiss_index_search.py:255]: [nitpicky] can we fix the indentation here?

Comment (emmafgrimes) [inline fairywren/clip/faiss_index_search.py:255]: Should this be `indexes`?

File: fairywren/clip/faiss_index_search.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+
+        test_embedding = fairywren_index.extract_clip_embedding(
+            test_preprocessed_multi_text
+            )
+
+        # DO PCA HERE or dimensionality
+        pcaed_embedding = pca(test_embedding)
+
+        fairywren_index.search(pcaed_embedding)
+        ```
```
Comment (emmafgrimes) [inline fairywren/clip/faiss_index_search.py:318]: Should we add a parameters section to this docstring?

File: tests/fixtures/data_fixtures.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+
+@pytest.fixture()
+def test_embeddings_non512():
+    return np.ones((2, 4), dtype=np.float32)
+
+
+@pytest.fixture()
+def test_metadata():
+
```
Comment (emmafgrimes) [inline tests/fixtures/data_fixtures.py:20]: Can we remove this line break to be consistent with above? Same for L26 below

File: tests/fixtures/data_fixtures.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+def test_metadata_mismatch():
+
+    return ['url1']
+
+
+@pytest.fixture()
+def fairywren_clip_index():
+    wren_index = CLIPFaissIndex(embedding_array=np.ones((2, 512), dtype=np.float32),
+                                id_metadata=['url1', 'url2']
+                                )
```
Comment (emmafgrimes) [inline tests/fixtures/data_fixtures.py:34]: [totally optional] mostly curious, would this work?
```suggestion
def fairywren_clip_index(test_embeddings, test_metadata):
    wren_index = CLIPFaissIndex(embedding_array=test_embeddings,
                                id_metadata=test_metadata
                                )
```

File: tests/fixtures/data_fixtures.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+def non_dict_input():
+    return 'hello world'
+
+
+@pytest.fixture()
+def bad_input_dict():
+    return {'foo': 'hello world'}
+
+
+# unsure if this will run correctly on git TBD
```
Comment (emmafgrimes) [inline tests/fixtures/data_fixtures.py:48]: lol did it work?

File: tests/test_clip_index.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+                       )
+    assert ('Length of embeddings 2 must match' in str(e.value)) is True
+
+
+def test_non_dict_preprocessing(fairywren_clip_index, non_dict_input):
+
+    with pytest.raises(Exception) as e:
+
+        fairywren_clip_index.preprocess_input(non_dict_input)
+    print(str(e.value))
```
Comment (emmafgrimes) [inline tests/test_clip_index.py:45]: Do we need this print statement?

File: tests/test_clip_index.py
Comment (emmafgrimes) [inline tests/test_clip_index.py:1]: [totally optional] I like naming my test files based on my function files, so this would be `test_faiss_index_search.py` and the fixture file would be `faiss_index_search_fixtures.py`. This is a personal opinion and I do not feel strongly, just wanted to share.

**Review by emmafgrimes — CHANGES_REQUESTED** [review]
This is awesome @sugi-chan ! great work on this. I added a bunch of nitpicky, optional, and typo comments, but the code itself looks great!

File: fairywren/clip/faiss_index_search.py [hunk 12]
```diff
@@ -0,0 +1,343 @@
+import faiss
+import numpy as np
+import torch
+from transformers import CLIPModel, CLIPProcessor
+
+
+base_clip_model = CLIPModel.from_pretrained('openai/clip-vit-base-patch32')
+base_processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
```
[the remark this answers is not in the document]
Comment (sugi-chan) [reply fairywren/clip/faiss_index_search.py:8]: im actually a monster

File: tests/fixtures/data_fixtures.py (hunk 10 above)
Comment (sugi-chan) [reply tests/fixtures/data_fixtures.py:48]: surprisingly yes, i forget what case is where it doesnt actually work... we did something different in octopod but I forget why

File: tests/test_clip_index.py (hunk 11 above)
Comment (sugi-chan) [reply tests/test_clip_index.py:45]: lol no. This was for me when I was crying about tests failing

File: tests/fixtures/data_fixtures.py (hunk 9 above)
Comment (sugi-chan) [reply tests/fixtures/data_fixtures.py:34]: yeah I think that technically would have worrrkeeeddddddd????

File: fairywren/clip/__init__.py [hunk 13]
```diff
@@ -0,0 +1 @@
+from fairywren.clip.faiss_index_search import CLIPFaissIndex
```
[the remark this answers is not in the document]
Comment (sugi-chan) [reply fairywren/clip/__init__.py:1]: classic

File: fairywren/clip/faiss_index_search.py [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+    def _build_index(self):
+        """
+        Builds a Faiss index based on `embedding_array` provided when initializing CLIPFaissIndex
+
+        For more information check https://github.com/facebookresearch/faiss
+
+        NOTES
+        -----
+        Faiss assumes a uniform structure of embeddings so the number of dimensions `d`
+        is calculated bas on the length of the first element of `self.embedding_array`
```
[the remark this answers is not in the document]
Comment (emmafgrimes) [reply fairywren/clip/faiss_index_search.py]: same!

File: fairywren/clip/faiss_index_search.py (hunk 4 above)
Comment (sugi-chan) [reply fairywren/clip/faiss_index_search.py:206]: things my brain didnt even register while looking through the code... gogo reviews

File: setup.py [hunk 15]
```diff
@@ -8,18 +8,24 @@
     readme = r.read()
 
 setup(
-    name='ltk_fairywren',
+    name='fairywren',
```
[the remark this answers is not in the document]
Comment (sugi-chan) [reply setup.py:11]: yeah I figured it was something of that sort. I did check ahead of time and fairywren is available on pypi haha. i guess it could be problematic if it gets taken by something. unsure

[the remark this answers is not in the document]
Comment (sugi-chan) [reply setup.py:11]: so fairywren is a potential candidate for open sourcing depending on how we continue to expand it. i dont have real preferences around this

[the remark this answers is not in the document]
Comment (emmafgrimes) [reply setup.py:11]: makes sense to leave as is!  just wanted to call out why we added the `ltk_` in `beaver`.

File: fairywren/clip/faiss_index_search.py [hunk 16]
```diff
@@ -316,6 +319,24 @@ def standard_search(self, clip_inputs, result_count=10):
...
+        result_count: int
+            number of results to return, defaults to 10
+
+        Returns
+        -------
+            list of dictionaries with 3 keys. `metadata`, `dist`, and `indexes`
+                metadata: contains metadata for each item returned by the faiss index search
+                based on the metadata parameter when the CLIPIndex class is built
+                dist: The reported distance between seed embedding and results.
+                indexes: the index numbers of results from the faiss index search.
```
Comment (emmafgrimes) [inline fairywren/clip/faiss_index_search.py:339]: one last funky indent ;)

**Review by emmafgrimes — APPROVED** [review]
Amazing! Thanks for making all the little changes, this is very exciting!

**Review by emmafgrimes — APPROVED** [review]
Not sure if you were waiting on me for this 🚢
